contracts/Delegations.sol

```diff
@@ -148,56 +159,84 @@ contract Delegations is IDelegations, IStakeChangeNotifier, ContractRegistryAcce
 	function stakeMigration(address _stakeOwner, uint256 _amount) external onlyStakingContract {}
 
 	function _stakeChangeBatch(address[] memory _stakeOwners, uint256[] memory _amounts, bool[] memory _signs) private returns (address[] memory delegates){
-		uint256[] memory newUncappedStakes = new uint256[](_stakeOwners.length);
-		uint256[] memory prevGovStakeOwners = new uint256[](_stakeOwners.length);
-		uint256[] memory prevGovStakeDelegates = new uint256[](_stakeOwners.length);
+		if (_stakeOwners.length == 0) {
+			return delegates;
+		}
 		delegates = new address[](_stakeOwners.length);
 
+		uint sequenceCount = 0;
+		address currentSeqDelegate = address(0);
+		for (uint i = 0; i < _stakeOwners.length; i++) { // count sequences and gather addresses
+			delegates[i] = getDelegation(_stakeOwners[i]);
+			if (currentSeqDelegate != delegates[i]) { // init record new delegate, and close previous one
+				sequenceCount++;
+				currentSeqDelegate == delegates[i];
+			}
+		}
+
+		uint256[] memory prevUncappedStakes = new uint256[](sequenceCount);
+		uint256[] memory newUncappedStakes = new uint256[](sequenceCount);
+		bool[] memory isSelfDelegatingDelegates = new bool[](sequenceCount);
+		address[] memory seqDelegates = new address[](sequenceCount);
+
+		// second pass
+		sequenceCount = 0;
+		currentSeqDelegate = address(0);
+		uint currentUncappedStake = 0;
 		for (uint i = 0; i < _stakeOwners.length; i++) {
-			(newUncappedStakes[i], prevGovStakeOwners[i], delegates[i], prevGovStakeDelegates[i]) = _applyStakeChangeLocally(_stakeOwners[i], _amounts[i], _signs[i]);
+			if (currentSeqDelegate != delegates[i]) {
+				if (sequenceCount > 0) { // close prev seq
+					uncappedStakes[currentSeqDelegate] = currentUncappedStake;
+					newUncappedStakes[sequenceCount - 1] = currentUncappedStake;
+				}
+				// init next seq
+				currentSeqDelegate = delegates[i];
+				currentUncappedStake = uncappedStakes[currentSeqDelegate];
+				seqDelegates[sequenceCount] = currentSeqDelegate;
+				prevUncappedStakes[sequenceCount];
```

Comment:
after fixing assignment - add test to check totalGovernanceStake is correct after batch staking change